Allow Deletion of Empty Workspace Groups

cailafinn opened the pull request:
**Description of work.**
Allows the deleteWorkspaces algorithm used in the GUI for the Delete key and button to work on empty groups. There is a check in Algorithm for empty groups that throws if it finds one. Rather than adding checks in  deleteWorkspace, which could have greater reaching consequesnses, we now simply make use of the UngroupWorkspace algorithm if the base class throws a specific error. 

**To test:**
Run the following
```python
from mantid import api
group = api.WorkspaceGroup()
mtd.addOrReplace("group", group)
```

Try to delete this by using the delete key or GUI button. It should be removed without errors. 

Fixes #33221

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

Commit:
Fix being unable to delete empty group workspaces

RE #33221

Commit:
Improve method design

RE #33221

Commit:
Add test for new functionality

RE #33221

Commit:
Add release notes

RE #33221

martyngigg commented on Framework/Algorithms/src/DeleteWorkspaces.cpp:
```diff
@@ -35,11 +35,27 @@ void DeleteWorkspaces::exec() {
-      if (!deleteAlg->isExecuted() || !success) {
+      bool success = false;
+      bool executed = false;
+      try {
```
Do we know why `DeleteWorkspace` can't just handle this?

ConorMFinn replied:
There's a call to a very long method in `Algorithm`, `checkGroups()`, which thows the exception. It's the running of the `DeleteWorkspace` child algorithm that causes the throw. It could be overridden in `DeleteWorkspace` but it seemed that would be too much of a change for the release at the time

Commit:
Generalise fix for groups

Original solution wouldn't have fixed the problem with trying to run a
singlular DeleteWorkspace. By overriding checkGroups and adding a
pre-check we ensure that empty groups can be deleted and the existing
behaviour is maintained.

RE #33221

DavidFair commented on Framework/Algorithms/src/DeleteWorkspace.cpp:
```diff
@@ -26,4 +30,15 @@ void DeleteWorkspace::exec() {
 }
+
+bool DeleteWorkspace::checkGroups() {
+  AnalysisDataServiceImpl &dataStore = AnalysisDataService::Instance();
```
Could we put a comment explaining why we override here

DavidFair replied:
Also I like this solution better, it nicely encapsulates everything too :+1:

Commit:
Fix failing tests for workspaces without names

RE #33221

Commit:
Add clarifying comment

RE #33221

DavidFair reviewed (dismissed):
:+1: this works for me with Ungroup, Delete, Clear and calling `DeleteWorkspace` as an alg too

martyngigg commented on Framework/Algorithms/src/DeleteWorkspaces.cpp:
```diff
@@ -7,6 +7,7 @@
 #include "MantidAlgorithms/DeleteWorkspaces.h"
 #include "MantidAPI/ADSValidator.h"
 #include "MantidAPI/AnalysisDataService.h"
+#include "MantidAPI/WorkspaceGroup.h"
```
Could we remove this unused include? Accumulating these plagues our build times...

martyngigg reviewed (changes requested):
Thanks for this nice fix. It's nice and now it doesn't need to touch algorithm I think it's safe for the release given the tests we have.

There's just an extra include we no longer need and can I ask that we also add a test in `DeleteWorkspaceTest`? As the actual behavioural change happens there then I think we need to test that too.

Commit:
Remove unused includes

RE #33221

Commit:
Add extra test for the algorithm that was actually changed

RE #33221

martyngigg reviewed (approved):
Thanks for addressing those comments.
